File: app/src/main/java/de/xikolo/controllers/main/SplashActivity.java
package de.xikolo.controllers.main;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;

import org.jetbrains.annotations.NotNull;

import java.util.Date;
import java.util.concurrent.TimeUnit;

import de.xikolo.controllers.dialogs.ApiVersionDeprecatedDialog;
import de.xikolo.controllers.dialogs.ApiVersionDeprecatedDialogAutoBundle;
import de.xikolo.controllers.dialogs.ApiVersionExpiredDialog;
import de.xikolo.controllers.dialogs.ServerErrorDialog;
import de.xikolo.controllers.dialogs.ServerMaintenanceDialog;
import de.xikolo.jobs.CheckHealthJob;
import de.xikolo.jobs.base.RequestJobCallback;

public class SplashActivity extends AppCompatActivity {

    public static final String TAG = SplashActivity.class.getSimpleName();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        new CheckHealthJob(getHealthCheckCallback()).run();
    }

    private RequestJobCallback getHealthCheckCallback() {
        return new RequestJobCallback() {
            @Override
            protected void onSuccess() {
                startApp();
            }

            @Override
            protected void onError(@NotNull ErrorCode code) {
                switch (code) {
                    case NO_NETWORK:
                        startApp();
                        break;
                    case API_VERSION_EXPIRED:
                        showApiVersionExpiredDialog();
                        break;
                    case MAINTENANCE:
                        showServerMaintenanceDialog();
                        break;
                    case ERROR:
                        showServerErrorDialog();
                        break;
                }
            }

            @Override
            protected void onDeprecated(@NonNull Date deprecationDate) {
                Date now = new Date();
                long distance = deprecationDate.getTime() - now.getTime();
                long days = TimeUnit.DAYS.convert(distance, TimeUnit.MILLISECONDS);

                if (days <= 14) {
                    showApiVersionDeprecatedDialog(deprecationDate);
                } else {
                    startApp();
                }
            }
        };
    }

    private void showApiVersionExpiredDialog() {
        ApiVersionExpiredDialog dialog = new ApiVersionExpiredDialog();
        dialog.setDialogListener(new ApiVersionExpiredDialog.ApiVersionExpiredDialogListener() {
            @Override
            public void onOpenPlayStoreClicked() {
                openPlayStore();
            }

            @Override
            public void onDismissed() {
                closeApp();
            }
        });
        showDialog(dialog, ApiVersionExpiredDialog.TAG);
    }

    private void showApiVersionDeprecatedDialog(Date deprecationDate) {
        ApiVersionDeprecatedDialog dialog = ApiVersionDeprecatedDialogAutoBundle.builder(deprecationDate).build();
        dialog.setDialogListener(new ApiVersionDeprecatedDialog.ApiVersionDeprecatedDialogListener() {
            @Override
            public void onOpenPlayStoreClicked() {
                openPlayStore();
            }

            @Override
            public void onDismissed() {
                startApp();
            }
        });
        showDialog(dialog, ApiVersionDeprecatedDialog.TAG);
    }

    private void showServerMaintenanceDialog() {
        ServerMaintenanceDialog dialog = new ServerMaintenanceDialog();
        dialog.setDialogListener(this::closeApp);
        showDialog(dialog, ServerMaintenanceDialog.TAG);
    }

    private void showServerErrorDialog() {
        ServerErrorDialog dialog = new ServerErrorDialog();
        dialog.setDialogListener(this::closeApp);
        showDialog(dialog, ServerErrorDialog.TAG);
    }

    private void startApp() {
        Intent intent = new Intent(SplashActivity.this, MainActivity.class);
        startActivity(intent);
        finish();
    }

    private void closeApp() {
        finish();
    }

    private void openPlayStore() {
        final String appPackageName = getPackageName();
        try {
            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + appPackageName)));
        } catch (android.content.ActivityNotFoundException anfe) {
            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=" + appPackageName)));
        }
        finish();
    }

    // bug fix workaround
    private void showDialog(DialogFragment dialogFragment, String tag) {
        FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
        ft.add(dialogFragment, tag);
        ft.commitAllowingStateLoss();
    }

}

File: app/src/main/java/de/xikolo/controllers/main/SplashActivity.kt
package de.xikolo.controllers.main

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.Lifecycle
import de.xikolo.R
import de.xikolo.controllers.dialogs.*
import de.xikolo.extensions.observe
import de.xikolo.models.Storage
import de.xikolo.network.jobs.CheckHealthJob
import de.xikolo.network.jobs.base.HealthCheckNetworkState
import de.xikolo.network.jobs.base.HealthCheckNetworkStateLiveData
import de.xikolo.network.jobs.base.NetworkCode
import de.xikolo.storages.ApplicationPreferences
import de.xikolo.utils.extensions.createIfNotExists
import de.xikolo.utils.extensions.fileCount
import de.xikolo.utils.extensions.internalStorage
import de.xikolo.utils.extensions.publicAppStorageFolder
import java.io.File
import java.util.*
import java.util.concurrent.TimeUnit

class SplashActivity : AppCompatActivity() {

    companion object {
        val TAG: String = SplashActivity::class.java.simpleName
    }

    private val networkState: HealthCheckNetworkStateLiveData by lazy {
        HealthCheckNetworkStateLiveData()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        networkState
            .observe(this) {
                if (it.code == NetworkCode.STARTED) {
                    return@observe
                }

                when (it.code) {
                    NetworkCode.SUCCESS                -> startApp()
                    NetworkCode.NO_NETWORK             -> startApp()
                    NetworkCode.API_VERSION_EXPIRED    -> showApiVersionExpiredDialog()
                    NetworkCode.MAINTENANCE            -> showServerMaintenanceDialog()
                    NetworkCode.API_VERSION_DEPRECATED -> {
                        it as HealthCheckNetworkState

                        it.deprecationDate?.let { deprecationDate ->
                            val now = Date()
                            val distance = deprecationDate.time - now.time
                            val days = TimeUnit.DAYS.convert(distance, TimeUnit.MILLISECONDS)

                            if (days <= 30) {
                                showApiVersionDeprecatedDialog(deprecationDate)
                            } else {
                                startApp()
                            }
                        } ?: run {
                            startApp()
                        }
                    }
                    else                               -> showServerErrorDialog()
                }
            }

        migrateStorage()
    }

    private fun runHealthJob() {
        CheckHealthJob(networkState, false).run()
    }

    private fun migrateStorage() {
        if (!ApplicationPreferences().contains(getString(R.string.preference_storage))) {
            val old = Storage(publicAppStorageFolder)
            val new = Storage(File(internalStorage.file.absolutePath + File.separator + "Courses"))
            new.file.createIfNotExists()

            val fileCount = old.file.fileCount

            val progressDialog = ProgressDialogHorizontalAutoBundle.builder()
                .title(getString(R.string.app_name))
                .message(getString(R.string.dialog_app_being_prepared))
                .build()
            progressDialog.max = 100
            progressDialog.show(supportFragmentManager, ProgressDialogHorizontal.TAG)

            old.migrateTo(new, object : Storage.MigrationCallback {
                override fun onProgressChanged(count: Int) {
                    runOnUiThread { progressDialog.progress = Math.ceil(100.0 * count / fileCount).toInt() }
                }

                override fun onCompleted(success: Boolean) {
                    runOnUiThread {
                        old.clean()
                        new.clean()
                        progressDialog.dismiss()
                        ApplicationPreferences().setToDefault(getString(R.string.preference_storage), getString(R.string.settings_default_value_storage))
                        runHealthJob()
                    }
                }
            })
        } else {
            runHealthJob()
        }
    }

    private fun showApiVersionExpiredDialog() {
        val dialog = ApiVersionExpiredDialog()
        dialog.listener = object : ApiVersionExpiredDialog.Listener {
            override fun onOpenPlayStoreClicked() {
                openPlayStore()
            }

            override fun onDismissed() {
                closeApp()
            }
        }
        showDialog(dialog, ApiVersionExpiredDialog.TAG)
    }

    private fun showApiVersionDeprecatedDialog(deprecationDate: Date) {
        val dialog = ApiVersionDeprecatedDialogAutoBundle.builder(deprecationDate).build()
        dialog.listener = object : ApiVersionDeprecatedDialog.Listener {
            override fun onOpenPlayStoreClicked() {}

            override fun onDismissed() {
                startApp()
            }
        }
        showDialog(dialog, ApiVersionDeprecatedDialog.TAG)
    }

    private fun showServerMaintenanceDialog() {
        val dialog = ServerMaintenanceDialog()
        dialog.listener = object : ServerMaintenanceDialog.Listener {
            override fun onDismissed() {
                closeApp()
            }
        }
        showDialog(dialog, ServerMaintenanceDialog.TAG)
    }

    private fun showServerErrorDialog() {
        val dialog = ServerErrorDialog()
        dialog.listener = object : ServerErrorDialog.Listener {
            override fun onDismissed() {
                closeApp()
            }
        }
        showDialog(dialog, ServerErrorDialog.TAG)
    }

    private fun startApp() {
        if (lifecycle.currentState.isAtLeast(Lifecycle.State.RESUMED)) {
            val intent = Intent(this@SplashActivity, MainActivity::class.java)
            startActivity(intent)
        }
        finish()
    }

    private fun closeApp() {
        finish()
    }

    private fun openPlayStore() {
        val appPackageName = packageName
        try {
            startActivity(
                Intent(
                    Intent.ACTION_VIEW,
                    Uri.parse("market://details?id=$appPackageName")
                )
            )
        } catch (e: android.content.ActivityNotFoundException) {
            startActivity(
                Intent(
                    Intent.ACTION_VIEW,
                    Uri.parse("https://play.google.com/store/apps/details?id=$appPackageName")
                )
            )
        }

        finish()
    }

    // bug fix workaround
    private fun showDialog(dialogFragment: DialogFragment, tag: String) {
        val ft = supportFragmentManager.beginTransaction()
        ft.add(dialogFragment, tag)
        ft.commitAllowingStateLoss()
    }

}
